### PR TITLE
Bug fix: allow empty `Vlines` and `Hlines` as `ggplot2` does

### LINKED ISSUE #1947: `ggplotly` fails with an empty layer of `geom_vline`
`ggplotly` fails when there is a `geom_vline`-layer that does not contain any lines.  For me, this happens in a setting where I have a function that computes a set of x-values at which I want to have vertical lines.  Sometimes it happens that no vertical line is needed, and then `ggplotly` fails with the following error message:

> Fehler in fix.by(by.y, y) : 'by' must specify a uniquely valid column

Minimal example:
```r
do_plot <- function(xintercepts) {
  ggplot(mtcars, aes(mpg, carb)) + geom_point() +
    geom_vline(xintercept = xintercepts)
}
plotly::ggplotly(do_plot(23))        # works fine
plotly::ggplotly(do_plot(c(15, 25))) # works fine
plotly::ggplotly(do_plot(numeric())) # produces error (but works without call to ggplotly)
```

### PR DESCRIPTION
**Pull request by dereckdemezquita**
Vlines and Hlines might sometimes be inexistent; `ggplot2` allows this, but `plotly` fails on merging empty `data.frame`.

I believe my solution is appropriate. I tested it locally, but if someone can help me through the process of writing a formal unit test, I would appreciate it.

Close #1947 

```r
# prep data
data <- data.table::data.table(gapminder::gapminder)
# add year of interest for vertical lines
data[year == 2002, year_of_interest := "yoi"]

# normal vertical line set manually at year 2000
ggplot2::ggplot(data, ggplot2::aes(x = year, y = lifeExp, colour = country)) +
    ggplot2::geom_line() +
    ggplot2::geom_vline(xintercept = 2000, colour = "red") +
    ggplot2::theme(legend.position = "none")

# vertical line by feeding data to it; this allows for programmatically setting many lines at different years
p <- ggplot2::ggplot(data, ggplot2::aes(x = year, y = lifeExp, colour = country)) +
    ggplot2::geom_line() +
    ggplot2::geom_vline(
        ggplot2::aes(xintercept = year),
        data = data[year_of_interest == "yoi", ],
        colour = "yellow",
        alpha = 0.75,
        linewidth = 1.25,
        linetype = "dashed"
    ) +
    ggplot2::theme(legend.position = "none")


# ggplot2 accepts that no lines are found plot is created and no error is thrown
p <- ggplot2::ggplot(data, ggplot2::aes(x = year, y = lifeExp, colour = country)) +
    ggplot2::geom_line() +
    ggplot2::geom_vline(
        ggplot2::aes(xintercept = year),
        data = data[year_of_interest == "some_not_found", ],
        colour = "yellow",
        alpha = 0.75,
        linewidth = 1.25,
        linetype = "dashed"
    ) +
    ggplot2::theme(legend.position = "none")

print(p)

# if we pass to plotly, it throws an error
plotly::ggplotly(p)
```

### CODE REVIEW

File: R/layers2traces.R [hunk 1]
```diff
@@ -445,7 +445,9 @@ to_basic.GeomHline <- function(data, prestats_data, layout, params, p, ...) {
     data = layout$layout, cols = paste0(x, c("_min", "_max")), values_to = x, names_to = "variable"
   ) 
   lay <- as.data.frame(lay)
-  data <- merge(lay[c("PANEL", x)], data, by = "PANEL")
+  if (nrow(data) != 0) {
```
Comment (cpsievert) [inline R/layers2traces.R:448]:
```suggestion
  if (nrow(data) > 0) {
```

File: R/layers2traces.R [hunk 2]
```diff
@@ -462,7 +464,10 @@ to_basic.GeomVline <- function(data, prestats_data, layout, params, p, ...) {
     data = layout$layout, cols = paste0(y, c("_min", "_max")), values_to = y, names_to = "variable"
   ) 
   lay <- as.data.frame(lay)
-  data <- merge(lay[c("PANEL", y)], data, by = "PANEL")
+  # fix for #1947; applied to Hline as well
+  if (nrow(data) != 0) {
```
Comment (cpsievert) [inline R/layers2traces.R:468]:
```suggestion
  if (nrow(data) > 0) {
```

File: tests/testthat/test-geom-vline-hline-allow-none-1947.R [hunk 3]
```diff
@@ -0,0 +1,77 @@
+
+test_that("geom_vline/geom_hline does not throw an error with ggplotly when no lines are found", {
```
Comment (cpsievert) [inline tests/testthat/test-geom-vline-hline-allow-none-1947.R:2]: It'd be great if these tests:

- Were added to `testthat/test-ggplot-hline.R`
- Were more minimal (i.e., no need to use datatable or gapminder... please use minimal datasets like the other tests in `testthat/test-ggplot-hline.R`

Comment (dereckdemezquita) [reply tests/testthat/test-geom-vline-hline-allow-none-1947.R:2]: Thanks @cpsievert I'll work on these updates later today :)

File: tests/testthat/test-geom-vline-hline-allow-none-1947.R [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      colour = "yellow",
+      alpha = 0.75,
+      linewidth = 1.25,
+      linetype = "dashed"
+    ) +
+    ggplot2::theme(legend.position = "none")
+
+  # error:
+  # "Error in fix.by(by.y, y) : 'by' must specify a uniquely valid column"
+  expect_error(plotly::ggplotly(p2), "Error in fix.by") # no lines are found no error is thrown
```
Comment (cpsievert) [inline tests/testthat/test-geom-vline-hline-allow-none-1947.R:39]: Isn't the point of the change here to make sure this _doesn't_ throw an error?

File: tests/testthat/test-geom-vline-hline-allow-none-1947.R (hunk 3 above)
Comment (dereckdemezquita) [reply tests/testthat/test-geom-vline-hline-allow-none-1947.R]: I am testing both hline and vline; I noticed there is a test-ggplot-vline.R file.

I simplified the data and corrected the expected results in the test assertion.

For now I've moved the test to the hline file as you suggested - I am expecting you to suggest separating these two into their respective files.

File: tests/testthat/test-geom-vline-hline-allow-none-1947.R (hunk 4 above)
Comment (dereckdemezquita) [reply tests/testthat/test-geom-vline-hline-allow-none-1947.R]: Corrected to `expect_error(plotly::ggplotly(p2), NA)`.

File: tests/testthat/test-geom-vline-hline-allow-none-1947.R (hunk 3 above)
Comment (dereckdemezquita) [reply tests/testthat/test-geom-vline-hline-allow-none-1947.R]: Hi @cpsievert you want me to make any more updates on this pull request? 

I separated the tests into their respective files; vertical/horizontal.

**Comment by dereckdemezquita** [comment]
@cpsievert does this PR still require more work? Let's get it merged?

File: tests/testthat/test-ggplot-hline.R [hunk 5]
```diff
@@ -97,3 +97,38 @@ test_that("hline works with coord_flip", {
   expect_equivalent(l$data[[2]]$x, c(5, 5))
   expect_equivalent(l$data[[2]]$y, c(5.95, 6.05))
 })
+
+# fix for issue #1974 pull request #2252
+test_that("geom_vline/geom_hline does not throw an error with ggplotly when no lines are found", {
+  set.seed(123)
```
Comment (cpsievert) [inline tests/testthat/test-ggplot-hline.R:103]:
```suggestion
```

File: tests/testthat/test-ggplot-vline.R [hunk 6]
```diff
@@ -46,3 +46,40 @@ test_that("vline works with coord_flip", {
   expect_equivalent(l$data[[2]]$x, c(5.95, 6.05))
   expect_equivalent(l$data[[2]]$y, c(5, 5))
 })
+
+
+# fix for issue #1974 pull request #2252
+test_that("geom_vline/geom_hline does not throw an error with ggplotly when no lines are found", {
+  set.seed(123)
```
Comment (cpsievert) [inline tests/testthat/test-ggplot-vline.R:53]:
```suggestion
```

File: tests/testthat/test-ggplot-hline.R [hunk 7]
```diff
@@ -97,3 +97,37 @@ test_that("hline works with coord_flip", {
...
+    ggplot2::geom_hline(
+      ggplot2::aes(yintercept = x),
+      data = df[df$value_of_interest == "something_not_matched", ],
+      colour = "pink"
+    )
+
+  # error given without fix:
+  # "Error in fix.by(by.y, y) : 'by' must specify a uniquely valid column"
+  expect_error(plotly::ggplotly(p4), NA) # no lines are found no error is thrown
+})
```
Comment (cpsievert) [inline tests/testthat/test-ggplot-hline.R:133]: Please make a similar edit to the vline test

```suggestion
test_that("geom_vline/geom_hline does not throw an error with ggplotly when no lines are found", {
  p3 <- ggplot(df) + geom_hline(aes(yintercept = x), data = data.frame(x = 1)[F, , drop = FALSE])
  expect_error(plotly::ggplotly(p4), NA) # no lines are found no error is thrown
})
```

Comment (dereckdemezquita) [reply tests/testthat/test-ggplot-hline.R]: I just did the one for `vline` as well.

```r
test_that("geom_vline/geom_hline does not throw an error with ggplotly when no lines are found", {
  p <- ggplot(df) + geom_vline(aes(xintercept = x), data = data.frame(x = 1)[F, , drop = FALSE])
  expect_error(plotly::ggplotly(p), NA)
})
```

**Comment by cpsievert** [comment]
Thank you, this is looking good, will you please also update the NEWS?

File: R/layers2traces.R [hunk 8]
```diff
@@ -462,7 +464,10 @@ to_basic.GeomVline <- function(data, prestats_data, layout, params, p, ...) {
     data = layout$layout, cols = paste0(y, c("_min", "_max")), values_to = y, names_to = "variable"
   ) 
   lay <- as.data.frame(lay)
-  data <- merge(lay[c("PANEL", y)], data, by = "PANEL")
+  # fix for #1947; applied to Hline as well
```
Comment (cpsievert) [inline R/layers2traces.R:467]:
```suggestion
```

**Pull request merged**